solidity/contracts/converter/BancorConverter.sol

```diff
@@ -528,44 +520,45 @@ contract BancorConverter is IBancorConverter, TokenHandler, SmartTokenController
       * @param _sourceToken source ERC20 token
       * @param _targetToken target ERC20 token
       * @param _amount      amount of tokens to convert (in units of the source token)
-      * @param _minReturn   if the conversion results in an amount smaller than the minimum return - it is cancelled, must be nonzero
       * @param _beneficiary wallet to receive the conversion result
       *
       * @return amount of tokens received (in units of the target token)
     */
-    function convertInternal(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount, uint256 _minReturn, address _beneficiary)
+    function convertInternal(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount, address _trader, address _beneficiary)
         public
         payable
         protected
         only(BANCOR_NETWORK)
-        greaterThanZero(_minReturn)
         returns (uint256)
     {
         require(_sourceToken != _targetToken); // validate input
 
+        // if a whitelist is set, verify that both and trader and the beneficiary are whitelisted
+        require(conversionWhitelist == address(0) ||
+                (conversionWhitelist.isWhitelisted(_trader) && conversionWhitelist.isWhitelisted(_beneficiary)));
+
         if (_targetToken == token)
-            return buy(_sourceToken, _amount, _minReturn, _beneficiary);
+            return buy(_sourceToken, _amount, _beneficiary);
         else if (_sourceToken == token)
-            return sell(_targetToken, _amount, _minReturn, _beneficiary);
+            return sell(_targetToken, _amount, _beneficiary);
         else
-            return crossConvert(_sourceToken, _targetToken, _amount, _minReturn, _beneficiary);
+            return crossConvert(_sourceToken, _targetToken, _amount, _beneficiary);
     }
 
     /**
       * @dev buys the smart token by depositing one of its reserve tokens
       * 
       * @param _reserveToken    reserve token contract address
       * @param _depositAmount   amount of tokens to deposit (in units of the reserve token)
-      * @param _minReturn       if the conversion results in an amount smaller than the minimum return - it is cancelled, must be nonzero
       * @param _beneficiary     wallet to receive the conversion result
       * 
       * @return amount of tokens received (in units of the smart token)
     */
-    function buy(IERC20Token _reserveToken, uint256 _depositAmount, uint256 _minReturn, address _beneficiary) internal returns (uint256) {
+    function buy(IERC20Token _reserveToken, uint256 _depositAmount, address _beneficiary) internal returns (uint256) {
         (uint256 amount, uint256 feeAmount) = getPurchaseReturn(_reserveToken, _depositAmount);
 
-        // ensure the trade gives something in return and meets the minimum requested amount
-        require(amount != 0 && amount >= _minReturn);
+        // ensure the trade gives something in return
+        require(amount != 0);

```

Comment:
@yudilevi : Doesn't this mean that we don't actually need to verify `minReturn > 0` everywhere?

Reply:
we do that to save some gas before doing the conversion in case min return is 0